### ISSUE TITLE
[Feature] Define DeepSeek V4 packed prefill metadata contract for var-len group batching

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto-lib#382]
### Summary

Define a DeepSeek V4 packed prefill kernel contract that can serve as the single-request (`num_reqs = 1`) and future vLLM-style var-len group-batch contract.

The goal is to refactor current fixed rectangular prefill interfaces away from `B`, `S`, and `T = B * S` assumptions, and toward a token-major execution model:

- runtime `num_reqs`
- runtime `num_tokens`
- token-major Q / hidden / MoE buffers
- per-request sequence metadata
- paged sliding-window KV metadata
- paged compressed KV metadata for CSA/HCA layers
- active-token MoE execution over `num_tokens`, not padded `B * S`

This issue is a concrete parameter-contract follow-up to #378.

### Motivation / Use Case

DeepSeek V4 prefill needs to move toward vLLM/SGLang-style batching, where each scheduler step builds an execution batch from active tokens rather than a rectangular `[batch, max_seq]` prompt tensor.

The current implementation still has many fixed-shape assumptions:

- `models/deepseek/v4/prefill_sparse_attn.py` uses standalone `B = 16`, `S = 128`, and `T = B * S`.
- Current attention code frequently derives request and sequence position from `t // S` and `t % S`.
- `seqused_kv[b]` bounds valid sequence length for correctness, but buffers and loop ranges still cover the rectangular `T`.
- DeepSeek V4 MoE (`moe.py`, `gate.py`, `dispatch.py`, `expert_shared.py`, `expert_routed.py`, `combine.py`) currently operates over all `T = B * S` tokens.
- DeepSeek V4 attention is layer-dependent: SWA uses only sliding-window KV, while CSA/HCA need compressed KV metadata and compressed sparse indices.

A B=1 first implementation is useful only if it is treated as the degenerate case of the final group-batch contract:

```text
num_reqs = 1
num_tokens = q_lens[0]
cu_q_lens = [0, num_tokens]
token_to_request[:] = 0
position_ids = start_pos + arange(num_tokens)
```

It should not become a separate `[1, S]` static path.

### Proposed API / Behavior

Introduce a packed DeepSeek V4 prefill contract with compile-time capacity and runtime active sizes:

```text
MAX_REQS       compile-time maximum request/group count
MAX_TOKENS     compile-time maximum active token count
num_reqs       runtime request/group count
num_tokens     runtime active token count
```

Core runtime metadata:

```text
num_reqs: pl.Scalar[pl.INT32]
num_tokens: pl.Scalar[pl.INT32]

q_lens: pl.Tensor[[MAX_REQS], pl.INT32]
cu_q_lens: pl.Tensor[[MAX_REQS + 1], pl.INT32]
context_lens: pl.Tensor[[MAX_REQS], pl.INT32]
kv_seq_lens: pl.Tensor[[MAX_REQS], pl.INT32]

token_to_request: pl.Tensor[[MAX_TOKENS], pl.INT32]
position_ids: pl.Tensor[[MAX_TOKENS], pl.INT32]
```

Parameter meanings:

- `num_reqs`: number of requests/groups in the current scheduler batch.
- `num_tokens`: total active prefill tokens in the current batch, equal to `sum(q_lens[:num_reqs])`.
- `q_lens[r]`: number of active prefill/suffix tokens for request `r` in this step.
- `cu_q_lens`: prefix sum over `q_lens`; request `r` occupies packed token range `[cu_q_lens[r], cu_q_lens[r + 1])`.
- `context_lens[r]`: KV length already present before this prefill step. This supports prefix-cache suffix prefill.
- `kv_seq_lens[r]`: total visible KV length after this prefill step, usually `context_lens[r] + q_lens[r]`.
- `token_to_request[t]`: request index for packed token `t`.
- `position_ids[t]`: absolute model position for packed token `t`; RoPE and compression boundaries must use this rather than `t % S`.

Token-major tensors should use `MAX_TOKENS` capacity and `num_tokens` validity:

```text
q: pl.Tensor[[MAX_TOKENS, H, HEAD_DIM], pl.BF16]
attn_out: pl.Out[pl.Tensor[[MAX_TOKENS, D], pl.BF16]]
input_ids: pl.Tensor[[MAX_TOKENS], pl.INT64]
hidden / x_mixed / x_norm: [MAX_TOKENS, D]
```

The kernel and orchestration should avoid:

```python
b = t // S
s = t % S
```

and use:

```python
r = token_to_request[t]
pos = position_ids[t]
q_begin = cu_q_lens[r]
q_end = cu_q_lens[r + 1]
ctx_len = context_lens[r]
kv_len = kv_seq_lens[r]
```

#### Common to all attention layer types

```text
attn_sink: pl.Tensor[[H], pl.FP32]

freqs_cos: pl.Tensor[[MAX_POSITION, ROPE_DIM], pl.BF16]
freqs_sin: pl.Tensor[[MAX_POSITION, ROPE_DIM], pl.BF16]
even_select_local: pl.Tensor[[ROPE_INTERLEAVE_CHUNK, ROPE_CHUNK], pl.BF16]
odd_select_local: pl.Tensor[[ROPE_INTERLEAVE_CHUNK, ROPE_CHUNK], pl.BF16]

wo_a: pl.Tensor[[O_GROUPS, O_LORA, O_GROUP_IN], pl.BF16]
wo_b: pl.Tensor[[D, O_GROUPS * O_LORA], pl.INT8]
wo_b_scale: pl.Tensor[[D], pl.FP32]
```

Meanings:

- `attn_sink`: per-head attention sink term.
- `freqs_cos` / `freqs_sin`: global RoPE tables indexed by `position_ids[t]`.
- `even_select_local` / `odd_select_local`: existing RoPE interleave helper tables.
- `wo_a`, `wo_b`, `wo_b_scale`: grouped output projection weights, unchanged from the current DeepSeek V4 sparse attention path.

#### Sliding-window/original KV metadata

```text
ori_slot_mapping: pl.Tensor[[MAX_TOKENS], pl.INT32]
ori_block_table: pl.Tensor[[MAX_REQS, ORI_MAX_BLOCKS], pl.INT32]
ori_kv: pl.Tensor[[ORI_BLOCK_NUM, BLOCK_SIZE, 1, HEAD_DIM], pl.BF16]
```

Meanings:

- `ori_slot_mapping[t]`: physical sliding-window/original KV slot for packed token `t`.
- `ori_block_table[r, logical_block]`: maps request-local original/sliding-window logical block to physical block.
- `ori_kv`: paged original/sliding-window KV storage.

#### `compress_ratio == 0`: SWA layers

SWA layers only need the sliding-window/original KV path. They should not require meaningful compressed KV writes.

Required metadata:

```text
ori_slot_mapping[MAX_TOKENS]
ori_block_table[MAX_REQS, ORI_MAX_BLOCKS]
ori_kv[ORI_BLOCK_NUM, BLOCK_SIZE, 1, HEAD_DIM]
window_sparse_indices[MAX_TOKENS, WIN] or cmp_sparse_indices[MAX_TOKENS, SPARSE_TOPK]
```

Behavior:

- Attention keys come from the per-request sliding window.
- The valid window for token `t` is derived from `token_to_request[t]`, `position_ids[t]`, `context_lens[r]`, and `kv_seq_lens[r]`.
- If using the existing `cmp_sparse_indices` shape for compatibility, compressed positions should be padded/invalid for SWA.
- No compressed KV state update is required.

#### `compress_ratio == 4`: CSA layers

CSA layers need sliding-window KV plus compressed KV metadata. They may require overlap-aware compressed writes and learned/indexer sparse selection depending on the final CSA path.

Required metadata:

```text
ori_slot_mapping[MAX_TOKENS]
ori_block_table[MAX_REQS, ORI_MAX_BLOCKS]
ori_kv[ORI_BLOCK_NUM, BLOCK_SIZE, 1, HEAD_DIM]

cmp_block_table[MAX_REQS, CMP_MAX_BLOCKS]
cmp_kv[CMP_BLOCK_NUM, BLOCK_SIZE, 1, HEAD_DIM]
cmp_sparse_indices[MAX_TOKENS, SPARSE_TOPK]

num_cmp_writes: pl.Scalar[pl.INT32]
cmp_write_token_ids[MAX_CMP_WRITES]
cmp_slot_mapping[MAX_CMP_WRITES]
```

Behavior:

- `cmp_sparse_indices[t, :]` is token-major and should be interpreted within request boundaries.
- Entries covering compressed KV must use `cmp_block_table[token_to_request[t], ...]`.
- `num_cmp_writes`, `cmp_write_token_ids`, and `cmp_slot_mapping` describe which active tokens produce compressed KV writes in this step.
- Compression boundaries should use `position_ids[t]`, not a batch-shared scalar `start_pos`.
- For B=1 suffix prefill, this is the same contract with `num_reqs = 1`.

#### `compress_ratio == 128`: HCA layers

HCA layers also need compressed KV metadata, but the compressed sparse portion may be generated deterministically rather than through a learned indexer. The same packed metadata should still be used so the contract remains uniform.

Required metadata:

```text
ori_slot_mapping[MAX_TOKENS]
ori_block_table[MAX_REQS, ORI_MAX_BLOCKS]
ori_kv[ORI_BLOCK_NUM, BLOCK_SIZE, 1, HEAD_DIM]

cmp_block_table[MAX_REQS, CMP_MAX_BLOCKS]
cmp_kv[CMP_BLOCK_NUM, BLOCK_SIZE, 1, HEAD_DIM]
cmp_sparse_indices[MAX_TOKENS, SPARSE_TOPK]

num_cmp_writes: pl.Scalar[pl.INT32]
cmp_write_token_ids[MAX_CMP_WRITES]
cmp_slot_mapping[MAX_CMP_WRITES]

cmp_kv_state[MAX_REQS, MAIN_STATE_LEN, MAIN_OUT_DIM]
cmp_score_state[MAX_REQS, MAIN_STATE_LEN, MAIN_OUT_DIM]
```

Behavior:

- HCA compressor state is request-scoped, so `cmp_kv_state` and `cmp_score_state` must be indexed by request metadata, not by rectangular batch index.
- Compression write decisions should be derived from per-token `position_ids` and request-local state.
- `cmp_sparse_indices` should remain `[MAX_TOKENS, SPARSE_TOPK]`, with only the first `num_tokens` rows valid.
- Deterministic compressed top-k generation should respect request boundaries and `kv_seq_lens[r]`.

### MoE Active-Token Contract

DeepSeek V4 packed prefill should also define the MoE path over active tokens:

```text
x_mixed: [MAX_TOKENS, D]
x_norm: [MAX_TOKENS, D]
x_norm_i8: [MAX_TOKENS, D]
x_norm_scale: [MAX_TOKENS, 1]
indices: [MAX_TOKENS, TOPK]
weights: [MAX_TOKENS, TOPK]
```

Required behavior:

- Gate, shared expert, dispatch, routed experts, and combine operate over `num_tokens`.
- `recv_token` stores packed token ids in `[0, num_tokens)`.
- `combine` produces packed `ffn_out[MAX_TOKENS, D]`.
- No routed expert traffic should be generated for padded capacity rows `t >= num_tokens`.

### Validation Plan

Add staged correctness checks:

1. B=1 full prefill:
   - `num_reqs = 1`
   - `context_lens = 0`
   - `q_lens = prompt_len`
   - compare packed output against existing full prefill for valid positions.

2. B=1 suffix prefill:
   - `num_reqs = 1`
   - `context_lens > 0`
   - `q_lens = suffix_len`
   - `position_ids = context_lens + arange(suffix_len)`
   - compare full prefill vs prefix-hit + suffix prefill final logits/KV.

3. Heterogeneous group batch:
   - multiple requests with different `q_lens` and `context_lens`
   - verify outputs match padded reference on valid token positions only.

4. SWA / CSA / HCA coverage:
   - SWA: `compress_ratio == 0`, no compressed writes.
   - CSA: `compress_ratio == 4`, compressed metadata and sparse indices valid.
   - HCA: `compress_ratio == 128`, request-scoped compressor state and deterministic compressed sparse indices.

5. MoE efficiency:
   - verify routed-token count scales with `num_tokens`, not `MAX_REQS * max_seq_len` or `B * S`.

### Alternatives Considered

- Keep the current rectangular `[B, S]` prefill and use `seqused_kv` only for masking. This preserves correctness but wastes QKV/projection/HC/MoE compute on padding.
- Add a special B=1 prefill path. This is useful only if it is implemented as `num_reqs = 1` under the packed contract; a separate `[1, S]` contract would create another migration target.
- Optimize only attention while leaving MoE rectangular. This reduces some waste but does not meet the DeepSeek V4 serving goal because MoE routed expert traffic remains padding-dependent.

### Additional Context

Related issues:

- #378 tracks the overall vLLM-style DeepSeek V4 dynamic prefill batching direction.
- #347 tracks current DeepSeek V4 prefill bring-up.
- #351 tracks per-batch `start_pos` support for DeepSeek V4 decode compressor paths.

This issue focuses specifically on the packed prefill parameter contract and layer-specific metadata needed before implementing var-len group batching.
